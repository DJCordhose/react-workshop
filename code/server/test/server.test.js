const request = require("supertest");

const db = require("../src/db");
const app = require("../src/app")(db);

function expectError(res, code) {
  expect(res.status).toBe(code);
  expect(res.body.error).toBeDefined();
}

beforeEach(() => {
  db.initialize();
});

describe("GET", () => {
  test("GET /greetings should return all greetings", () => {
    return request(app)
      .get("/greetings")
      .then(response => {
        expect(response.status).toBe(200);
        expect(response.body).toMatchSnapshot();
      });
  });
  //
  test("GET /greetings/id should return a single greeting by id", () => {
    return request(app)
      .get("/greetings/2")
      .then(response => {
        expect(response.status).toBe(200);
        expect(response.body).toMatchSnapshot();
      });
  });
  //
  test("GET /greetings/id should return 404 for an unknown id", () => {
    return request(app)
      .get("/greetings/666")
      .then(response => {
        expectError(response, 404);
      });
  });
  //
  // });
  //
  describe("POST", () => {
    test("POST /greetings with new greeting should save it and return new id", () => {
      return request(app)
        .post("/greetings")
        .send({ name: "Moni", greeting: "Huhu!" })
        .then(res => {
          expect(res.status).toBe(201);
          expect(res.body).toMatchSnapshot();
          const newGreetingInDatabase = db._greetings[10];
          expect(newGreetingInDatabase).toEqual({
            id: 11,
            name: "Moni",
            greeting: "Huhu!"
          });
        });
    });

    test("POST /greetings with empty body  should fail with HTTP 400", () => {
      return request(app)
        .post("/greetings")
        .then(res => {
          expectError(res, 400);
        });
    });

    test("POST /greetings with empty name should fail with HTTP 400", () => {
      return request(app)
        .post("/greetings")
        .send({ greeting: "Huhu!" })
        .then(res => expectError(res, 400));
    });

    test("POST /greetings with empty greeting should fail with HTTP 400", () => {
      return request(app)
        .post("/greetings")
        .send({ name: "Moni" })
        .then(res => expectError(res, 400));
    });
  });
});

describe("EventEmitter", () => {
  const listener = jest.fn();

  test("received notification on insert", done => {
    listener.mockImplementation(() => done());
    db.on("insert", listener);
    db.insert({ greeting: "Hello", name: "Listener" });
  });

  afterEach(() => {
    db.removeListener("insert", listener);
  });
});
