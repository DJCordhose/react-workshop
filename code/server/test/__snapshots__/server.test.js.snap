// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET GET /greetings should return all greetings 1`] = `
Array [
  Object {
    "greeting": "Moin",
    "id": 1,
    "name": "Klaus",
  },
  Object {
    "greeting": "Hello!",
    "id": 2,
    "name": "Susi",
  },
  Object {
    "greeting": "Bonjour",
    "id": 3,
    "name": "Max",
  },
  Object {
    "greeting": "How are you?",
    "id": 4,
    "name": "Susi",
  },
  Object {
    "greeting": "Bon soir",
    "id": 5,
    "name": "Max",
  },
  Object {
    "greeting": "Hola, ¿qué tal?",
    "id": 6,
    "name": "Felipe",
  },
  Object {
    "greeting": "Happy Birthday",
    "id": 7,
    "name": "Alex",
  },
  Object {
    "greeting": "¡buenos días",
    "id": 8,
    "name": "Felipe",
  },
  Object {
    "greeting": "Wie gehts?",
    "id": 9,
    "name": "Paul",
  },
  Object {
    "greeting": "Have a nice day",
    "id": 10,
    "name": "Susi",
  },
]
`;

exports[`GET GET /greetings/id should return a single greeting by id 1`] = `
Object {
  "greeting": "Hello!",
  "id": 2,
  "name": "Susi",
}
`;

exports[`GET POST POST /greetings with new greeting should save it and return new id 1`] = `
Object {
  "id": 11,
}
`;
